### DISCUSSION TITLE
classes/class_astar2d

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#257] [User-contributed notes]
# classes/class_astar2d

Inherits: RefCounted< Object An implementation of A* for finding the shortest path between two vertices on a connected graph in 2D space. Description: An implementation of the A* algorithm, used to...

https://docs.godotengine.org/en/stable/classes/class_astar2d.html

**Comment by amysthat** [comment]
For anyone wondering - `AStar2D` is not a node, and you won't be able to find it in the "Create New Node" window in the editor. It's a `RefCounted`, and it's only instantiated with the following code:
```gdscript
var astar2d = AStar2D.new()
```
In-engine documentation is available. It took my some good amount of time to understand this, so I thought it would be a good idea to share it here.

**Reply by WldGeese** [reply]
Thank you for this <3

**Comment by spikerXneh** [comment]
Is there an example of how this is used somewhere?

**Reply by Calinou** [reply]
Yes, check out this official demo which uses AStarGrid2D (a subclass of AStar2D constrained to grids): https://godotengine.org/asset-library/asset/2723

Pull request opened:

- https://github.com/godotengine/godot/pull/107069

**Comment by marcusViniciusMP** [comment]
im really cracking my head trying to gess why this class hates diagonal straight moves in my hexagonal tilemaplayer
must provide a working model teaching how to use it to make moves without zigzags, thanks !

**Reply by JamSpasm** [reply]
I am playing around with adding AStar pathfinding to a hexagonal tilemap right now and ran into some strange quirks with hexagon maps and coordinate systems. If you're getting movement but it is zig-zaggy double check that your points are in the right spot, and that each point is properly connected to it's neighbor. Hexagons find their neighbors differently depending if the Y value is even or odd. I found this resource useful:

https://mortoray.com/finding-neighbours-in-a-hex-map/#:~:text=To%20find%20the%20neighbours%20of,the%20number%20around%20the%20perimeter

**Reply by Squint-Eastwood** [reply]
Make sure you're not mixing this up with the variant: AStarGrid2D. I imagine that would cause problems with diagonal movement.
